src/Abstractions/NexusMods.Abstractions.Loadouts.Synchronizers/ALoadoutSynchronizer.cs

```diff
@@ -1183,81 +1183,96 @@ await Parallel.ForEachAsync(files, async (item, _) =>
     /// </summary>
     public async Task<bool> ReindexState(GameInstallation installation, IConnection connection, ITransaction tx)
     {
-        var seen = new HashSet<GamePath>();
-        var metadata = GameInstallMetadata.Load(connection.Db, installation.GameMetadataId);
-        var inState = metadata.DiskStateEntries.ToDictionary(e => (GamePath)e.Path);
-        var changes = false;
-        
-        
         var hashDb = await _fileHashService.GetFileHashesDb();
-        
-        foreach (var location in installation.LocationsRegister.GetTopLevelLocations())
+
+        var gameInstallMetadata = GameInstallMetadata.Load(connection.Db, installation.GameMetadataId);
+
+        var previousDiskStateEntities = gameInstallMetadata.DiskStateEntries;
+        var previousDiskState = new Dictionary<GamePath, DiskStateEntry.ReadOnly>(capacity: previousDiskStateEntities.Count);
+
+        foreach (var previousDiskStateEntity in previousDiskStateEntities)
         {
-            if (!location.Value.DirectoryExists())
-                continue;
+            GamePath path = previousDiskStateEntity.Path;
+
+            ref var diskStateEntity = ref CollectionsMarshal.GetValueRefOrAddDefault(previousDiskState, path, out var hasExistingDiskStateEntity);
+            if (hasExistingDiskStateEntity)
+            {
+                _logger.LogWarning("Duplicate path in disk state: `{Path}`", path);
+            }
+
+            diskStateEntity = previousDiskStateEntity;
+        }
 
-            await Parallel.ForEachAsync(location.Value.EnumerateFiles(), async (file, token) =>
+        var hasDiskStateChanged = false;
+
+        var seenPaths = new HashSet<GamePath>();
+        var seenPathsLock = new Lock();
+
+        foreach (var locationPair in installation.LocationsRegister.GetTopLevelLocations())
+        {
+            var (_, locationPath) = locationPair;
+            if (!locationPath.DirectoryExists()) continue;
+
+            await Parallel.ForEachAsync(locationPath.EnumerateFiles(), async (file, token) =>
+            {
+                var gamePath = installation.LocationsRegister.ToGamePath(file);
+                if (ShouldIgnorePathWhenIndexing(gamePath)) return;
+
+                bool isNewPath;
+                lock (seenPathsLock)
                 {
-                    {
-                        var gamePath = installation.LocationsRegister.ToGamePath(file);
-                        if (ShouldIgnorePathWhenIndexing(gamePath)) return;
+                    isNewPath = seenPaths.Add(gamePath);
```

Comment:
I guess in the case of a case-sensitive file system we could have the same path twice with different casing. 
Hopefully we don't run into that.

Reply:
We noted that we wouldn't support such a case:

https://github.com/Nexus-Mods/NexusMods.App/blob/main/docs/developers/decisions/backend/0003-paths.md